docs/design-documents/hal/0003-explicit-pinmap-extension.md

```diff
@@ -0,0 +1,302 @@
+# HAL Explicit PinMap Extension Design Document
+
+# Table of contents
+
+1. [HAL Explicit PinMap Extension Design Document](#hal-explicit-pinmap-extension-design-document).
+1. [Table of contents](#table-of-contents).
+1. [Introduction](#introduction).
+    1. [Overview and background](#overview-and-background).
+    1. [Requirements and assumptions](#requirements-and-assumptions).
+1. [System architecture and high-level design](#system-architecture-and-high-level-design).
+    1. [Explicit pinmap mechanism](#explicit-pinmap-mechanism).
+    1. [How much we can save](#how-much-we-can-save).
+1. [Detailed design](#detailed-design).
+    1. [Pinmap types](#pinmap-types).
+    1. [HAL API changes](#hal-api-changes).
+    1. [Drivers API changes](#drivers-api-changes).
+    1. [`constexpr` utility functions](#constexpr-utility-functions).
+    1. [`Example usage](#constexpr-utility-functions).
+
+# Introduction
+
+Making use of the pinmap tables optional by allowing the peripheral configuration (pin/periheral/function) to be explicitly specified in the HAL API function.
+
+### Overview and background
+
+HAL APIs making use of pins take these pins in their constructor and use those pins to lookup which peripheral/function to use. The process of looking up the peripheral/function requires there to be a pinmap table which maps pins to peripherals/functions. This pinmap table takes up ROM which could be saved if the pinmap wasn't used.
+The goal is to provide additional HAL API/constructors which takes pinmap as a parameter where pin/peripheral/function is specified explicitly and there is no need to use the pinmap tables.
+This extension should give the following savings:
+- removed pinmap tables,
+- removed `hal\mbed_pinmap_common library` (required for pin lookup mechanism),
+- reduced HAL driver code.
+
+### Requirements and assumptions
+
+1. Provide types which will hold explicit pinmaps for peripherals(`PWM`, `AnalogIn`, `AnalogOut`, `SPI`, `I2C`, `UART`, `QSPI`, `CAN`).
+2. Provide `xxx_init_direct(xxx_t *obj, explicit_pinmap_t *)` functions to HAL API (these functions will not use pinmap tables).
+3. Provide additional constructors in drivers layer which will use the `xxx_init_direct(xxx_t *obj, explicit_pinmap_t*)` HAL functions.
+4. Provide default weak implementations of `xxx_init_direct(explicit_pinmap_t *)` functions. These functions will call standard `xxx_init(xxx_t *obj, PinName, ...)` function (backward compatibility for targets which do not support explicit pinmap mechanism).
+5. Provide `constexpr` utility functions to lookup for pin mapping in compile time (requires C++14).
+6. Initialize console using explicit pinmap mechanism, so `hal\mbed_pinmap_common library` is not needed and can be removed.
```

Comment:
https://github.com/ARMmbed/mbed-os/blob/ba7b4799f9a74797f0380226748bb1d985b358ec/platform/source/mbed_retarget.cpp#L261-L280

The serial console is initialized using standard API which uses pin-maps. To get full benefits of explicit pinmap extension we need to change this for targets which provide explicit pinmap extension and use new API. This will give two benefits:
- Serial pinmap tables will be removed from the image,
- Pinmap lib will be removed from the image:
`| hal\mbed_pinmap_common.o        |      0(-272) |   0(+0) |      0(+0) | `

Related change can fe found here:
https://github.com/ARMmbed/mbed-os/pull/11399/commits/44154f7ffbe7f02e50cbac2be6c8c4cb9774b436